### PR TITLE
nodeos min-initial-block-num option 📦

### PR DESCRIPTION
**Pull request by huangminghuang**
## Change Description

This PR adds `min-initial-block-num` option to chain plugin so that nodeos could refuse to load any state from prior to the specified block number

## Change Type
**Select *ONE*:**
- [ ] Documentation

- [ ] Stability bug fix

- [x] Other

- [ ] Other - special case

## Testing Changes
**Select *ANY* that apply:**
- [ ] New Tests

- [ ] Existing Tests

- [ ] Test Framework

- [ ] CI System

- [ ] Other

## Consensus Changes
- [ ] Consensus Changes

## API Changes
- [ ] API Changes

## Documentation Additions
- [x] Documentation Additions

### CODE REVIEW

File: libraries/chain/controller.cpp [hunk 1]
```diff
@@ -619,6 +619,9 @@ struct controller_impl {
    void init(std::function<bool()> check_shutdown, bool clean_startup) {
       uint32_t lib_num = (blog.head() ? blog.head()->block_num() : fork_db.root()->block_num);
 
+      EOS_ASSERT( lib_num >= this->conf.min_initial_block_num, misc_exception, "Controller head "
+      "at block number ${block_num}, which is smaller than the minimum required ${required}", ("block_num", fork_db.head()->block_num)("required",this->conf.min_initial_block_num) );
```
Comment (heifner) [inline libraries/chain/controller.cpp]: This reports and uses the term head when lib is used. Also the variable name and option description should be clear that this feature is for lib and not head.

**Review by heifner — CHANGES_REQUESTED** [review]
lib vs head comment.

**Review by nksanthosh — COMMENTED** [review]
Will the implementation work with all 4 options (irrespective of how state is loaded)

1. Replay
2. From genesis
3. Snapshot
4. Load binary state file

Has it been manually tested against all 4 options?

Either add automated tests for all four options if it is quick or we can prioritize and add automated tests as a separate PR immediately following this one.

**Comment by huangminghuang** [comment]
> Will the implementation work with all 4 options (irrespective of how state is loaded)
> 
> 1. Replay
> 2. From genesis
> 3. Snapshot
> 4. Load binary state file
> 
> Has it been manually tested against all 4 options?
> 
> Either add automated tests for all four options if it is quick or we can prioritize and add automated tests as a separate PR immediately following this one.

Unit tests are available to test against each scenario.

File: libraries/chain/controller.cpp [hunk 2]
```diff
@@ -619,8 +619,8 @@ struct controller_impl {
    void init(std::function<bool()> check_shutdown, bool clean_startup) {
       uint32_t lib_num = (blog.head() ? blog.head()->block_num() : fork_db.root()->block_num);
 
-      EOS_ASSERT( lib_num >= this->conf.min_initial_block_num, misc_exception, "Controller head "
-      "at block number ${block_num}, which is smaller than the minimum required ${required}", ("block_num", fork_db.head()->block_num)("required",this->conf.min_initial_block_num) );
+      EOS_ASSERT( lib_num >= this->conf.min_initial_block_num, misc_exception, "Controller latest irreversible block "
+      "at block number ${lib_num}, which is smaller than the minimum required ${required}", ("block_num", lib_num)("required",this->conf.min_initial_block_num) );
```
Comment (linhuang-blockone) [inline libraries/chain/controller.cpp]: "block_num" should be "lib_num"

File: libraries/chain/controller.cpp [hunk 3]
```diff
@@ -619,6 +619,9 @@ struct controller_impl {
    void init(std::function<bool()> check_shutdown, bool clean_startup) {
       uint32_t lib_num = (blog.head() ? blog.head()->block_num() : fork_db.root()->block_num);
 
+      EOS_ASSERT( lib_num >= this->conf.min_initial_block_num, misc_exception, "Controller latest irreversible block "
+      "at block number ${lib_num}, which is smaller than the minimum required ${required}", ("block_num", lib_num)("required",this->conf.min_initial_block_num) );
```
Comment (victorj8) [inline libraries/chain/controller.cpp]: I think this should be ${block_num} instead of ${lib_num}

File: unittests/restart_chain_tests.cpp [hunk 4]
```diff
@@ -205,6 +205,43 @@ BOOST_AUTO_TEST_CASE(test_existing_state_without_block_log) {
...
+   blocks.push_back(chain.produce_block());
+   blocks.push_back(chain.produce_block());
+   blocks.push_back(chain.produce_block());
+   chain.close();
+
+   auto cfg                  = chain.get_config();
+   cfg.min_initial_block_num = 100;
+   remove_existing_blocks(cfg);
+
+   // restarting chain with no block log and no genesis
```
Comment (linhuang-blockone) [inline unittests/restart_chain_tests.cpp:237]: Can you add a test for normal cases `lib == cfg.min_initial_block_num`  and `lib > cfg.min_initial_block_num`?

**Review by linhuang-blockone — APPROVED** [review]

File: plugins/chain_plugin/chain_plugin.cpp [hunk 5]
```diff
@@ -458,6 +458,8 @@ void chain_plugin::set_program_options(options_description& cli, options_descrip
          ("export-reversible-blocks", bpo::value<bfs::path>(),
            "export reversible block database in portable format into specified file and then exit")
          ("snapshot", bpo::value<bfs::path>(), "File to read Snapshot State from")
+         ("min-initial-block-num", bpo::value<uint32_t>()->default_value(0), "minimum latest irreversible block number to load, "
+         "nodeos would refuse to load any state from prior to that block number")
```
Comment (heifner) [inline plugins/chain_plugin/chain_plugin.cpp]: Should not use hard coded "nodeos". Either don't mention "nodeos" or use the executable name variable. I prefer just not mentioning it. Also I think it should be "will" not "would".

Maybe something like:
"minimum last irreversible block (lib) number, fail to start if state/snapshot lib is prior to specified"

**Review by heifner — APPROVED** [review]

**Review by nksanthosh — APPROVED** [review]

**Pull request merged**